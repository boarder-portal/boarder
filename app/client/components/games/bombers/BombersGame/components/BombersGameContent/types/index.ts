import { Images } from 'client/hooks/useImages';

export type BomberImage = 'grass' | 'wall' | 'box' | 'bomb' | 'bonusBomb' | 'bonusRange' | 'bonusSpeed' | 'bonusHp';

export type BombersImages = Images<BomberImage>;
