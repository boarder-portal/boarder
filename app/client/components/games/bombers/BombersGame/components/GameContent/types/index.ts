import { ImagesDictionary } from 'client/hooks/useImages';

export type BomberImage = 'grass' | 'wall' | 'box' | 'bomb' | 'bonusBomb' | 'bonusRange' | 'bonusSpeed' | 'bonusHp';

export type BombersImages = ImagesDictionary<BomberImage>;
