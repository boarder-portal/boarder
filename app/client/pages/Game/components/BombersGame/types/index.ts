import { TImagesDictionary } from 'client/hooks/useImages';

export type TBomberImage = 'grass' | 'wall' | 'box' | 'bomb' | 'bonusBomb' | 'bonusRange' | 'bonusSpeed' | 'bonusHp';

export type TBombersImages = TImagesDictionary<TBomberImage>;
