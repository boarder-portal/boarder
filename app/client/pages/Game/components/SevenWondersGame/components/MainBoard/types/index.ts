import {
  IResource,
  TAction,
  TPayments,
  TResourceOwner,
} from 'common/types/sevenWonders';

export interface IOwnerResource extends IResource {
  owner: TResourceOwner;
}

export interface ISevenWondersCourtesansBuildInfo {
  cardIndex: number;
  action: TAction;
  payments?: TPayments
}
