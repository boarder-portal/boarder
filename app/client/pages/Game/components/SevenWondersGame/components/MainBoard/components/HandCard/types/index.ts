import { TSevenWondersBuildType, TSevenWondersPayments } from 'common/types/sevenWonders';

import {
  ITradeVariant,
} from 'client/pages/Game/components/SevenWondersGame/components/MainBoard/components/HandCard/utilities/getTradeVariantsByPurchaseVariants';

export enum EBuildType {
  FREE = 'FREE',
  FREE_WITH_EFFECT = 'FREE_WITH_EFFECT',
  FREE_BY_BUILDING = 'FREE_BY_BUILDING',
  FREE_BY_OWN_RESOURCES = 'FREE_BY_OWN_RESOURCES',
  OWN_RESOURCES_AND_COINS = 'OWN_RESOURCES_AND_COINS',
  WITH_TRADE = 'WITH_TRADE',
  NOT_ENOUGH_RESOURCES_OR_COINS = 'NOT_ENOUGH_RESOURCES_OR_COINS',
  ALREADY_BUILT = 'ALREADY_BUILT',
  NOT_ALLOWED = 'NOT_ALLOWED',
}

export interface IBuildInfo {
  type: EBuildType;
  title: string;
  tradeVariants: ITradeVariant[];
  onBuild(freeBuildType: TSevenWondersBuildType | null, payments?: TSevenWondersPayments): void;
}
