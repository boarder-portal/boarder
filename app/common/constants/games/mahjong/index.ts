import invert from 'lodash/invert';
import times from 'lodash/times';

import { EDragon, EHandsCount, ESet, ESuit, EWind, IGameOptions, IPlayerSettings } from 'common/types/mahjong';

export const DEFAULT_GAME_OPTIONS: IGameOptions = {
  minPlayersCount: 4,
  maxPlayersCount: 4,
  handsCount: EHandsCount.FOUR,
};

export const DEFAULT_PLAYER_SETTINGS: IPlayerSettings = {
  autoPass: true,
  sortHand: false,
  showLosingHand: true,
  showCurrentTile: false,
};

export const ALL_SUITS = Object.values(ESuit);
export const ALL_VALUES = times(9, (x) => x + 1);
export const ALL_DRAGONS = Object.values(EDragon);
export const ALL_WINDS = Object.values(EWind);

export const FLOWERS_STRING_VALUES = '12345678';
export const SUITED_STRING_VALUES = '123456789';

export const DRAGON_TO_STRING_MAP: Record<EDragon, string> = {
  [EDragon.RED]: 'r',
  [EDragon.GREEN]: 'g',
  [EDragon.WHITE]: 'w',
};

export const STRING_TO_DRAGON_MAP = invert(DRAGON_TO_STRING_MAP) as Partial<Record<string, EDragon>>;

export const WIND_TO_STRING_MAP: Record<EWind, string> = {
  [EWind.EAST]: 'e',
  [EWind.SOUTH]: 's',
  [EWind.WEST]: 'w',
  [EWind.NORTH]: 'n',
};

export const STRING_TO_WIND_MAP = invert(WIND_TO_STRING_MAP) as Partial<Record<string, EWind>>;

export const SUIT_TO_STRING_MAP: Record<ESuit, string> = {
  [ESuit.BAMBOOS]: 'b',
  [ESuit.CHARACTERS]: 'c',
  [ESuit.DOTS]: 'd',
};

export const STRING_TO_SUIT_MAP = invert(SUIT_TO_STRING_MAP) as Partial<Record<string, ESuit>>;

export const SET_NAMES: Record<ESet, string> = {
  [ESet.PAIR]: 'Пара',
  [ESet.PUNG]: 'Панг',
  [ESet.KONG]: 'Конг',
  [ESet.CHOW]: 'Чоу',
  [ESet.KNITTED_CHOW]: 'Переплетенное чоу',
};

export const WIND_SHORT_NAMES: Record<EWind, string> = {
  [EWind.EAST]: 'E',
  [EWind.SOUTH]: 'S',
  [EWind.WEST]: 'W',
  [EWind.NORTH]: 'N',
};
