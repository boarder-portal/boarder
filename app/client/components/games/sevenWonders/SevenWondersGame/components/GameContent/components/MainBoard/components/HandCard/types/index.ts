import { BuildType, Payments } from 'common/types/games/sevenWonders';

import { TradeVariant } from 'client/components/games/sevenWonders/SevenWondersGame/components/GameContent/components/MainBoard/components/HandCard/utilities/getTradeVariantsByPurchaseVariants';

export enum BuildKind {
  FREE = 'FREE',
  FREE_WITH_EFFECT = 'FREE_WITH_EFFECT',
  FREE_BY_BUILDING = 'FREE_BY_BUILDING',
  FREE_BY_OWN_RESOURCES = 'FREE_BY_OWN_RESOURCES',
  OWN_RESOURCES_AND_COINS = 'OWN_RESOURCES_AND_COINS',
  WITH_TRADE = 'WITH_TRADE',
  NOT_ENOUGH_RESOURCES_OR_COINS = 'NOT_ENOUGH_RESOURCES_OR_COINS',
  ALREADY_BUILT = 'ALREADY_BUILT',
  NOT_ALLOWED = 'NOT_ALLOWED',
}

export interface BuildInfo {
  type: BuildKind;
  title: string;
  tradeVariants: TradeVariant[];
  onBuild(freeBuildType: BuildType | null, payments?: Payments): void;
}
