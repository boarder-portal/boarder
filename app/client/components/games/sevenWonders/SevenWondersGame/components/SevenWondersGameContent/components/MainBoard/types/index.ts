import { Action, Payments, Resource, ResourceOwner } from 'common/types/games/sevenWonders';

export interface OwnerResource extends Resource {
  owner: ResourceOwner;
}

export interface CourtesansBuildInfo {
  cardIndex: number;
  action: Action;
  payments?: Payments;
}
