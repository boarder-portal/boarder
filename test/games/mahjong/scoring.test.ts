import { EWind } from 'common/types/mahjong';

import { getHandMahjong } from 'common/utilities/mahjong/scoring';
import { parseTile, parseTiles } from 'common/utilities/mahjong/parse';
import { stringifyMahjong } from 'common/utilities/mahjong/stringify';

const standardOptions = {
  roundWind: EWind.EAST,
  seatWind: EWind.EAST,
  isSelfDraw: true,
  meldedSets: [],
  concealedSets: [],
  flowers: [],
  isLastTile: false,
  isRobbingKong: false,
  isReplacementTile: false,
  isLastWallTile: false,
};

describe('mahjong', () => {
  describe('scoring', () => {
    test('basic mahjong', () => {
      expect(
        stringifyMahjong(
          getHandMahjong({
            ...standardOptions,
            hand: parseTiles('WeWeWe WsWsWs WwWwWw WnWnWn Dr'),
            waits: parseTiles('Dr'),
            winningTile: parseTile('Dr'),
          }),
        ),
      ).toMatchSnapshot();
    });
  });
});
