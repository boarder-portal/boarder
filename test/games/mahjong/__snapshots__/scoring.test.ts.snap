// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`mahjong scoring basic mahjong 1`] = `
"221

BIG_FOUR_WINDS (pu We, pu Ws, pu Ww, pu Wn)
ALL_HONORS
FOUR_CONCEALED_PUNGS (pu We, pu Ws, pu Ww, pu Wn)
FULLY_CONCEALED_HAND
SINGLE_WAIT

pu We
pu Ws
pu Ww
pu Wn
pa Dr"
`;
